Vagrant Script: Detect colliding active virtualbox VMs and warn users

vsk-coding opened the pull request:
By running vagrant script when there is active residual VMs that are not
visible to vagrant creates error.

This commit fix that error by listing the active residual VMs and
presenting the user with options to Pause, Stop or Destroy the residual
VMs

...
 RUNTIME=containerd K8S=1 NWORKERS=1 contrib/vagrant/start.sh

 Detected running VMs that might cause conflict

 "cilium_k8s1_1609849557356_67827" {55360ba9-a7f6-41b5-ad8f-c600df308f14}
 "cilium_k8s2_1609851496114_28645" {e5b8239b-3762-4e98-9184-066375623e2b}

 Do you wish to pause (p/P), stop (s/S), or destroy (d/D)?
 ...

 Fixes: #13928

 Signed-off-by: Vishnu Soman K <vishnusomank05@gmail.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

Fixes: #issue-number

```release-note
<!-- Enter the release note text here if needed or remove this section! -->
```

vishnusomank committed:
Vagrant Script: Detect colliding active virtualbox VMs and warn users

By running vagrant script when there is active residual VMs that are not
visible to vagrant creates error.

This commit fix that error by listing the active residual VMs and
presenting the user with options to Pause, Stop or Destroy the residual
VMs

...
 RUNTIME=containerd K8S=1 NWORKERS=1 contrib/vagrant/start.sh

 Detected running VMs that might cause conflict

 "cilium_k8s1_1609849557356_67827" {55360ba9-a7f6-41b5-ad8f-c600df308f14}
 "cilium_k8s2_1609851496114_28645" {e5b8239b-3762-4e98-9184-066375623e2b}

 Do you wish to pause (p/P), stop (s/S), or destroy (d/D)?
 ...

 Fixes: #13928

 Signed-off-by: Vishnu Soman K <vishnusomank05@gmail.com>

pchaigno commented on contrib/vagrant/start.sh:
```diff
@@ -576,36 +576,92 @@ function vboxnet_addr_finder(){
+	echo
+	VBoxManage list runningvms
+	echo
+	printf "Do you wish to pause (p/P), stop (s/S), or destroy (d/D)?  "
```
I think we should also propose to simply ignore the issue. Especially considering the script cannot make a distinction between Cilium's VM and other VMs. So, with the current set of choices, a user with their own VM running would be blocked from running any Cilium VM.

pchaigno replied:
Capitalized letters should be reserved to the default choice, IMO "continue". So we would have something like:
```
Do you wish to pause, stop, or destroy the VM(s) or ignore and continue? [p/s/d/C]
```

pchaigno replied:
You mean per-VM? I don't think we need it per VM if that's what you're asking.

pchaigno commented on contrib/vagrant/start.sh:
```diff
@@ -576,36 +576,92 @@ function vboxnet_addr_finder(){
+				VBoxManage controlvm $VMName pause
+			done
+			printf "\n$runningVm VM(s) paused successfully\n" 
+			printf "Tip: Try running vagrant status to check status\n"
```
Pausing the VM is not currently a viable option because the script doesn't seem to distinguish between paused and running VMs:
```
$ NO_BUILD=1 K8S=1 ./contrib/vagrant/start.sh 

Detected running VMs that might cause conflict

"cilium3_k8s1_1609951352422_54263" {93a231c6-c3f9-4366-a20f-35f60623b33b}

Do you wish to pause (p/P), stop (s/S), or destroy (d/D)?  p

cilium3_k8s1_1609951352422_54263

1 VM(s) paused successfully
Tip: Try running vagrant status to check status
$
$
$ NO_BUILD=1 K8S=1 ./contrib/vagrant/start.sh 

Detected running VMs that might cause conflict

"cilium3_k8s1_1609951352422_54263" {93a231c6-c3f9-4366-a20f-35f60623b33b}

Do you wish to pause (p/P), stop (s/S), or destroy (d/D)?  p

cilium3_k8s1_1609951352422_54263
VBoxManage: error: Already paused
VBoxManage: error: Details: code VBOX_E_INVALID_VM_STATE (0x80bb0002), component ConsoleWrap, interface IConsole, callee nsISupports
VBoxManage: error: Context: "Pause()" at line 405 of file VBoxManageControlVM.cpp

1 VM(s) paused successfully
Tip: Try running vagrant status to check status
```

pchaigno commented on contrib/vagrant/start.sh:
```diff
@@ -576,36 +576,92 @@ function vboxnet_addr_finder(){
-fi
-
+	createVm
+fi
```
Nit: missing new line here.

pchaigno reviewed (changes requested):
Thanks for taking a stab at this @vsk-coding! IMO, it's the sort of small changes that can make a big difference when we onboard new contributors to Cilium :tada: 

I have a couple of comments below on the options we provide the users when they have existing VMs running.

pchaigno reviewed (changes requested):
Another small issue I missed initially:
We allow for the syntax:
```
K8S=1 NETNEXT=1 ./contrib/vagrant/start.sh
K8S=1 NETNEXT=1 NWORKERS=1 ./contrib/vagrant/start.sh k8s2+
```
to enable us to provision a first VM, then provision the second only later, when we need it. That behavior is broken with the current changes because it complains k8s1+ already exists when trying to create k8s2+.

vishnusomank committed:
Vagrant Script: Detect colliding active virtualbox VMs and warn users

By running vagrant script when there is active residual VMs that are not
visible to vagrant creates error.

This commit fix that error by listing the active residual VMs and
presenting the user with options to Pause, Stop or Destroy the residual
VMs or ignore and continue.

...
$NO_BUILD=1 K8S=1 ./contrib/vagrant/start.sh

Detected running VMs that might cause conflict

"cilium_k8s1_1610031382439_78313" {a3c160c3-9f2b-4aae-8c87-44e004a7a003}

Do you wish to pause, stop, or destroy the VM(s) or ignore and continue?
[p/s/d/C]
...

Fixes: #13928

Signed-off-by: Vishnu Soman K vishnusomank05@gmail.com

vishnusomank committed:
Vagrant Script: Detect colliding active virtualbox VMs and warn users

By running vagrant script when there is active residual VMs that are not
visible to vagrant creates error.

This commit fix that error by listing the active residual VMs and
presenting the user with options to Pause, Stop or Destroy the residual
VMs or ignore and continue.

...
$NO_BUILD=1 K8S=1 ./contrib/vagrant/start.sh

Detected running VMs that might cause conflict

"cilium_k8s1_1610031382439_78313" {a3c160c3-9f2b-4aae-8c87-44e004a7a003}

Do you wish to pause, stop, or destroy the VM(s) or ignore and continue?
[p/s/d/C]  C
...

Fixes: #13928

Signed-off-by: Vishnu Soman K <vishnusomank05@gmail.com>
